### PR TITLE
e2e: Add test for testing wallet name editing

### LINKED ISSUE #442: End-to-End Testing Coverage
## Overview

End-to-end testing is an essential part of ensuring Caravan works correctly from a user's perspective. Our e2e tests simulate real user workflows by interacting with the actual application interface, coordinating with a Bitcoin regtest node, and verifying that critical flows work as expected.

This is an excellent area for new contributors to get familiar with Caravan since it involves working with actual user flows without requiring deep knowledge of the internal architecture. Each test scenario helps you understand how Caravan works while improving our test coverage.

## How E2E Testing Works in Caravan

Our e2e tests use Playwright to automate browser interactions with the Caravan Coordinator application. When you run the tests, the system automatically:

1. Spins up a Bitcoin regtest node in Docker
2. Launches the Caravan application 
3. Creates test wallets with funded addresses
4. Runs automated tests that click buttons, fill forms, and verify outcomes
5. Cleans up all resources when finished

Everything runs in an isolated environment, so you can safely test transaction flows without touching real Bitcoin.

## Current Test Coverage

We currently have three main test suites covering core workflows:

### 01-multisig_setup.spec.ts
Tests the wallet creation and configuration flow:
- Loading the Caravan homepage
- Navigating through wallet setup
- Testing Bitcoin node connections with various scenarios (success, wrong URL, incorrect credentials)
- Creating a 2-of-3 multisig wallet
- Entering extended public keys from test wallets
- Downloading wallet configuration files

### 02-wallet_regtest.spec.ts
Tests wallet configuration and address management:
- Modifying wallet configurations for regtest network
- Extracting and setting proper BIP32 paths and fingerprints
- Importing wallets into Caravan
- Importing addresses to Bitcoin node
- Verifying address generation and display
- Address pagination and navigation

### 03-transaction_flow.spec.ts
Tests transaction creation and broadcasting:
- Creating transactions with automatic coin selection
- Creating transactions with manual UTXO selection
- Downloading unsigned PSBTs
- Signing PSBTs with multiple wallets
- Uploading signatures (both as text and file upload)
- Broadcasting transactions
- Verifying successful transaction broadcast

### Prerequisites
- Docker installed and running
- Node.js (check `.nvmrc` for version)

Also you'd need to change resource setting of Docker to allow atleast 8gbs of ram access

Image

### Getting Started

1. **Clone and navigate to the project:**
   ```bash
   git clone https://github.com/caravan-bitcoin/caravan.git
   cd caravan
   ```

2. **Install dependencies:**
   ```bash
   npm install
   cd apps/coordinator
   npm run test:e2e:install  # Installs Playwright browsers
   ```

3. **Set up environment:**
   ```bash
   cd e2e
   cp .env.example .env
   ```

4. **Run the tests:**
   ```bash
   # From project root
   npm run test:e2e --filter=caravan-coordinator
   
   # Or from coordinator directory
   cd apps/coordinator
   npm run test:e2e          # Run all tests
   npm run test:e2e:ui       # Interactive UI mode
   npm run test:e2e:headed   # See the browser
   ```

The test infrastructure handles everything automatically - no need to manually start Docker containers or configure the Bitcoin node.

### Reliability of existing tests

In addition to the below tests to build out, there is some flakiness that exists in the current tests where timeouts will eventually be hit. Our tests will run retries and if after 3 retries they still fail then the suite will fail. Recreating these failures and trying to improve the performance and reliability would be an important contribution to our codebase.

Here are workflows we'd like to add e2e coverage for. Each represents a small, focused PR opportunity:

### Wallet Management
- [ ] Importing an existing wallet configuration file
- [ ] Testing different address types (P2WSH, P2SH-P2WSH, P2TR)
- [ ] Verifying wallet with different quorum configurations (1-of-2, 3-of-5)
- [ ] Testing wallet name editing
- [ ] Exporting wallet descriptors

### Address Operations
- [ ] Generating and displaying change addresses
- [ ] Testing address verification with QR codes
- [ ] Batch address generation
- [ ] Address reuse warnings

### Transaction Workflows
- [ ] PSBT management (downloading, uploading, importing with signatures)
- [ ] Creating transactions to multiple recipients
- [ ] RBF (Replace-By-Fee) transaction creation
- [ ] CPFP (Child-Pays-For-Parent) transaction creation
- [ ] Testing different fee rate selections
- [ ] Transaction preview and review flow
- [ ] Canceling transaction creation mid-flow

### Error Handling
- [ ] Invalid address entry
- [ ] Insufficient balance scenarios
- [ ] Invalid PSBT upload
- [ ] Signature mismatch errors
- [ ] Network disconnection handling
- [ ] Invalid fee rate inputs

### Hardware Wallet Integration
- [ ] Coldcard PSBT import/export flow
- [ ] Testing different signing device combinations
- [ ] Verifying proper input ordering

### Client Connections
- [ ] Testing different public API providers
- [ ] Private node connection failure recovery
- [ ] Network switching (testnet/regtest)

### 1. Pick a Test Scenario
Choose one of the unchecked items from the list above, or propose a new one by commenting on this issue. 

IMPORTANT: Choose just one per PER to make them easier to review and validate.

### 2. Set Up Your Development Environment
Follow the setup instructions above to get e2e tests running locally.

### 3. Write Your Test
Create a new test file or add to an existing one in `apps/coordinator/e2e/tests/`. Look at the existing tests as examples. Each test should:
- Have a clear, descriptive name
- Focus on a single workflow or scenario
- Use the shared test utilities in `e2e/utils/` and `e2e/testhelpers/`
- Clean up after itself (handled by global teardown)

### 4. Test Locally
Run your test multiple times to ensure it's stable:
```bash
npm run test:e2e:headed  # Watch it run
npm run test:e2e         # Verify it passes consistently
```

### 5. Submit a Pull Request
Create a PR with:
- A descriptive title like "e2e: Add test for RBF transaction creation"
- A brief description of what the test covers
- Reference to this issue
- Any setup notes or edge cases discovered

## Example Test Structure

```typescript
test.describe("Your Test Suite Name", () => {
  test.beforeEach(async ({ page }) => {
    // Navigate to starting point
    await page.goto("/#/wallet");
    // Load wallet, connect to node, etc.
  });

  test("should perform the workflow", async ({ page }) => {
    // Your test steps here
    // Use page.locator() to find elements
    // Use expect() to verify outcomes
  });
});
```

## Helpful Resources

- **Playwright Documentation**: https://playwright.dev/docs/intro
- **Existing Test Examples**: `apps/coordinator/e2e/tests/`
- **Test Helpers**: `apps/coordinator/e2e/testhelpers/` - Utilities for common operations
- **Bitcoin RPC Client**: `apps/coordinator/e2e/utils/bitcoinClient.ts` - For blockchain operations

## Questions or Need Help?

Feel free to comment on this issue or join our discord channel or here on the issue . We're here to help new contributors get started with testing and learn about Caravan along the way :) 

---

*Note: A demo video showing how to run and write e2e tests is added below .* ( I did it for mac if anyone does it on windows/linux , please feel free to share here

### PR DESCRIPTION
**Pull request by Ferryx349**
**What kind of change does this PR introduce?**

- Test improvement / E2E test coverage

**Issue Number:**

Fixes #442 

**If relevant, did you update the documentation?**

- Not applicable (test-only change)

**Summary**

This PR adds an end-to-end test to cover the wallet name editing functionality in the Caravan coordinator.

Previously, editing and saving a wallet name was not covered by E2E tests. This change ensures that:
- A loaded wallet can enter edit mode
- The wallet name can be updated and saved
- The updated name is correctly reflected in the UI

The test follows the existing E2E test structure and error-handling conventions and runs as part of the full coordinator E2E suite.

**Does this PR introduce a breaking change?**

- No

## Checklist
- [x] I have tested my changes thoroughly.
- [x] I have added or updated tests to cover my changes (if applicable).
- [ ] I have verified that test coverage meets or exceeds 95% (if applicable).
- [x] I have run the test suite locally, and all tests pass.
- [x] I have written tests for all new changes/features
- [x] I have followed the project's coding style and conventions.
- [ ] I have created a changeset to document my changes (`npm run changeset`).

**Other information**

- The test handles Playwright strict mode by scoping locators where multiple elements share the same `data-cy` attributes.
- All coordinator E2E tests pass locally using Docker-based setup.

**Have you read the [contributing guide](https://github.com/caravan-bitcoin/caravan/blob/main/apps/coordinator/CONTRIBUTING.md)?**

- Yes

### CODE REVIEW

File: apps/coordinator/e2e/tests/02-wallet_regtest.spec.ts [hunk 1]
```diff
@@ -190,4 +190,43 @@ test.describe("Wallet Regtest Configuration", () => {
...
+    // Fill new name
+    const nameInput = page.getByLabel("Name");
+    await expect(nameInput).toBeVisible();
+    await nameInput.fill(newName);
+
+    // Save change
+    await page.locator('[data-cy="save-button"]').click();
+
+    // Verify updated name is rendered
+    await expect(nameDisplay).toHaveText(newName);
```
Comment (Legend101Zz) [inline apps/coordinator/e2e/tests/02-wallet_regtest.spec.ts]: I guess this test doesn't verify the name was actually saved to the backend. Consider:
```ts
// Reload page to verify persistence
await page.reload();
await expect(nameDisplay).toHaveText(newName);
```

also maybe should be verify that after saving,  the UI returns to display mode ?

**Review by Legend101Zz — CHANGES_REQUESTED** [review]
Thanks for the PR , overall looks good , just a couple of reviews

File: apps/coordinator/e2e/tests/02-wallet_regtest.spec.ts (hunk 1 above)
Comment (Legend101Zz) [reply apps/coordinator/e2e/tests/02-wallet_regtest.spec.ts]: no worries , I guess you can revert back to as it was originally done by you , sorry for giving you the trouble :)

**Comment by Abhay349** [comment]
Hi @Legend101Zz I have updated the test to address the race condition and kept it focused on validating the user-visible behavior only. The test now checks that the wallet name updates correctly in the UI and that the UI exits edit mode after saving.
I have also updated the inline comments to better explain the intent behind each assertion. 
Screenshot 2026-02-11 at 07 48 29
All tests are passing.
Let me know if there issues, happy to iterate on further feedback.
Thanks :)

**Review by Legend101Zz — APPROVED** [review]
@Abhay349 thanks for bearing with me and all the changes .. I am approving this PR for @bucko13 's review :)

File: apps/coordinator/e2e/tests/02-wallet_regtest.spec.ts [hunk 2]
```diff
@@ -190,4 +190,47 @@ test.describe("Wallet Regtest Configuration", () => {
...
     }
   });
+  test("should allow user to edit and save wallet name", async ({ page }) => {
+  try {
+    await page.goto("/#/wallet");
+    await expect(page).toHaveURL(/.*\/wallet/);
+
+    // Wallet name should be visible (display mode)
+    let nameDisplay = page.locator('[data-cy="editable-name-value"]').first();
+    await expect(nameDisplay).toBeVisible({ timeout: 15000 });
```
[the remark this answers is not in the document]
Comment (Abhay349) [reply apps/coordinator/e2e/tests/02-wallet_regtest.spec.ts]: And I'm happy to modify existing tests based on feedback.
Thanks :)

**Comment by bucko13** [comment]
e2e has been rewritten to be more robust but has introduced some conflicts and so this will need to be fixed.

**Comment by Legend101Zz** [comment]
@Abhay349 Thanks for your patience. As Buck suggested above due to the E2E changes the testsiote was updated , Could you please pull in the latest changes and review the updated E2E setup we’ve implemented, then adjust your changes accordingly?

**Comment by Ferryx349** [comment]
> @Abhay349 Thanks for your patience. As Buck suggested above due to the E2E changes the testsiote was updated , Could you please pull in the latest changes and review the updated E2E setup we’ve implemented, then adjust your changes accordingly?

on it

**Comment by Ferryx349** [comment]
Hi @bucko13 @Legend101Zz  I have fixed the conflicts, let me know if there are further issues, happy to iterate on further feedback :)
